ext/mixed/js/japanese.js

```diff
@@ -17,24 +17,153 @@
  */
 
 
-function jpIsKanji(c) {
-    const code = c.charCodeAt(0);
-    return code >= 0x4e00 && code < 0x9fb0 || code >= 0x3400 && code < 0x4dc0;
+const JP_HALFWIDTH_KATAKANA_MAPPING = new Map([
+    ['ｦ', 'ヲヺ-'],
+    ['ｧ', 'ァ--'],
+    ['ｨ', 'ィ--'],
+    ['ｩ', 'ゥ--'],
+    ['ｪ', 'ェ--'],
+    ['ｫ', 'ォ--'],
+    ['ｬ', 'ャ--'],
+    ['ｭ', 'ュ--'],
+    ['ｮ', 'ョ--'],
+    ['ｯ', 'ッ--'],
+    ['ｰ', 'ー--'],
+    ['ｱ', 'ア--'],
+    ['ｲ', 'イ--'],
+    ['ｳ', 'ウヴ-'],
+    ['ｴ', 'エ--'],
+    ['ｵ', 'オ--'],
+    ['ｶ', 'カガ-'],
+    ['ｷ', 'キギ-'],
+    ['ｸ', 'クグ-'],
+    ['ｹ', 'ケゲ-'],
+    ['ｺ', 'コゴ-'],
+    ['ｻ', 'サザ-'],
+    ['ｼ', 'シジ-'],
+    ['ｽ', 'スズ-'],
+    ['ｾ', 'セゼ-'],
+    ['ｿ', 'ソゾ-'],
+    ['ﾀ', 'タダ-'],
+    ['ﾁ', 'チヂ-'],
+    ['ﾂ', 'ツヅ-'],
+    ['ﾃ', 'テデ-'],
+    ['ﾄ', 'トド-'],
+    ['ﾅ', 'ナ--'],
+    ['ﾆ', 'ニ--'],
+    ['ﾇ', 'ヌ--'],
+    ['ﾈ', 'ネ--'],
+    ['ﾉ', 'ノ--'],
+    ['ﾊ', 'ハバパ'],
+    ['ﾋ', 'ヒビピ'],
+    ['ﾌ', 'フブプ'],
+    ['ﾍ', 'ヘベペ'],
+    ['ﾎ', 'ホボポ'],
+    ['ﾏ', 'マ--'],
+    ['ﾐ', 'ミ--'],
+    ['ﾑ', 'ム--'],
+    ['ﾒ', 'メ--'],
+    ['ﾓ', 'モ--'],
+    ['ﾔ', 'ヤ--'],
+    ['ﾕ', 'ユ--'],
+    ['ﾖ', 'ヨ--'],
+    ['ﾗ', 'ラ--'],
+    ['ﾘ', 'リ--'],
+    ['ﾙ', 'ル--'],
+    ['ﾚ', 'レ--'],
+    ['ﾛ', 'ロ--'],
+    ['ﾜ', 'ワ--'],
+    ['ﾝ', 'ン--']
+]);
+
+const JP_HIRAGANA_RANGE = [0x3040, 0x309f];
+const JP_KATAKANA_RANGE = [0x30a0, 0x30ff];
+const JP_KANA_RANGES = [JP_HIRAGANA_RANGE, JP_KATAKANA_RANGE];
+
+const JP_CJK_COMMON_RANGE = [0x4e00, 0x9fff];
+const JP_CJK_RARE_RANGE = [0x3400, 0x4dbf];
+const JP_CJK_RANGES = [JP_CJK_COMMON_RANGE, JP_CJK_RARE_RANGE];
+
+const JP_ITERATION_MARK_CHAR_CODE = 0x3005;
+
+// Japanese character ranges, roughly ordered in order of expected frequency
+const JP_JAPANESE_RANGES = [
+    JP_HIRAGANA_RANGE,
+    JP_KATAKANA_RANGE,
+
+    JP_CJK_COMMON_RANGE,
+    JP_CJK_RARE_RANGE,
+
+    [0xff66, 0xff9f], // Halfwidth katakana
+
+    [0x30fb, 0x30fc], // Katakana punctuation
+    [0xff61, 0xff65], // Kana punctuation
```

Comment:
Are these left out of `JP_KANA_RANGES` on purpose? It seems to be used in `jpIsStringEntirelyKana` which exists in `jpConvertReading` only. WanaKana's `toRomaji` doesn't seem to be able to handle these anyway, but this would work
```javascript
wanakana.toRomaji('か'+String.fromCharCode('0x3099'))
"ka゙"
wanakana.toRomaji(('か'+String.fromCharCode('0x3099')).normalize())
"ga"
```
These character codes differ between `jpIsCharCodeKana` and `wanakana.isKana`. Both ignore halfwidth katakana:
```
0x3040	぀
0x3097	゗
0x3098	゘
0x3099	゙
0x309a	゚
0x309b	゛
0x309c	゜
0x309d	ゝ
0x309e	ゞ
0x309f	ゟ
0x30a0	゠
0x30fd	ヽ
0x30fe	ヾ
0x30ff	ヿ
```

Reply:
I can just revert `jpIsStringEntirelyKana` to be `wanakana.isKana`, which is the previous behaviour. I'm not concerned about performance in that area. That functionality was outside of the scope of what I was trying to change in this commit, it just got caught up in the changes of what `jpIsKana` was being used for (both single char and multi char). Do you think that sounds like a good idea? If we need more changes such as the `normalize` thing, that could be done separately.

Reply:
The change shouldn't break anything unless some Yomichan or MeCab dictionaries use the combining marks instead of direct character codes for が、ぱ etc. The romaji ruby text is probably also quite an obscure feature. I'm fine if it's left like this for now.

I think it's good to be explicit about what the function does, because I was quite surprised when `jpIsKana` and `wanakana.isKana` returned false for a string that was only partially kana. Now the name tells the functionality even if `wanakana.isKana` is used.